### PR TITLE
Add-Tabledata + tests

### PR DESCRIPTION
**Pull request by ahandley**
Sorry in advance for the length of this commit.  It has a bunch of versioning hell fixes and the size is padded by the content files that I added for testing.  

The main files to look over are the test file and BqTabledata.cs.  Other than the length, they shouldn't be too dense.  (So many parameters!)

### CODE REVIEW

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1 [hunk 1]
```diff
@@ -93,4 +93,113 @@ Describe "Set-BqSchema" {
...
+        $table.NumRows | Should Be 20
+    }
+
+    It "should handle write disposition WriteIfEmpty" {
+        $table | Add-BqTabledata $filename_broken_csv CSV -SkipLeadingRows 1 -MaxBadRecords 4 `
+                -AllowUnknownFields -ErrorAction SilentlyContinue
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 7
+        { $table | Add-BqTabledata $filename_csv CSV -SkipLeadingRows 1 `
+            -WriteMode WriteIfEmpty -ErrorAction Stop } | Should Throw "404"
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1]: We should test scenario where user use -WriteIfEmpty and no error is thrown too. Same for -WriteTruncate and -WriteAppend.

File: Google.PowerShell.IntegrationTests/BigQuery/classics_broken.csv [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+War and Peas,1869
+Lime and Punishment,Fyodor Dostoyevsky,1866,extra
+Saffron and Sensibility,Jane Austin,1811,extra
+Moby Dill,1851
+Don Quinoa,Miguel de Cervantes,1605
+One Hundred Years of Salad,Gabriel Marquez,1967
+Ender's Garnish,1985
+Pride and Pomegranite,Jane Austin,1813
+A Tale of Two Ciders,Charles Dickens,1859
+Omelette,William Shakespeare,1603,extra
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/classics_broken.csv]: nit: add new line.

File: Google.PowerShell.Tests/packages.config [hunk 3]
```diff
@@ -1,15 +1,20 @@
...
+  <package id="Google.Cloud.BigQuery.V2" version="1.0.0-beta09" targetFramework="net452" />
   <package id="log4net" version="2.0.3" targetFramework="net452" />
   <package id="Newtonsoft.Json" version="7.0.1" targetFramework="net452" />
   <package id="NUnit" version="2.6.4" targetFramework="net452" />
   <package id="NUnitTestAdapter" version="2.0.0" targetFramework="net452" />
+  <package id="System.Interactive.Async" version="3.1.1" targetFramework="net452" />
+  <package id="System.Net.Http" version="4.3.0" targetFramework="net452" />
   <package id="Zlib.Portable.Signed" version="1.11.0" targetFramework="net452" />
-</packages>
+</packages>
```
Comment (quoctruong) [inline Google.PowerShell.Tests/packages.config]: Add new line here.

File: Google.PowerShell/BigQuery/BqTabledata.cs [hunk 4]
```diff
@@ -230,4 +231,206 @@ protected override void EndProcessing()
...
+        [Parameter(Mandatory = false)]
+        public SwitchParameter AllowUnknownFields { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// CSV ONLY: Allows insertion of rows that are missing trailing optional columns.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        public SwitchParameter AllowJaggedRows { get; set; }
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Do you have a test for this or AllowQuotedNewlines?

File: Google.PowerShell/BigQuery/BqTabledata.cs [hunk 5]
```diff
@@ -230,4 +231,206 @@ protected override void EndProcessing()
...
+        {
+            Stream fileInput;
+            try
+            {
+                fileInput = new StreamReader(Filename).BaseStream;
+            }
+            catch (Exception ex)
+            {
+                WriteError(new ErrorRecord(ex,
+                    $"Error while reading file {Filename}.",
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Put quote around fileName.

File: Google.PowerShell/BigQuery/BqTabledata.cs [hunk 6]
```diff
@@ -230,4 +231,206 @@ protected override void EndProcessing()
...
+            }
+            catch (Exception ex)
+            {
+                WriteError(new ErrorRecord(ex,
+                    $"Error while uploading file '{Filename}' to table '{InputObject.TableId}'.",
+                    ErrorCategory.WriteError, Filename));
+            }
+        }
+    }
+}
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: New line,

File: Google.PowerShell/packages.config [hunk 7]
```diff
@@ -1,24 +1,29 @@
...
   <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net452" />
   <package id="Microsoft.Bcl.Build" version="1.0.21" targetFramework="net452" />
   <package id="Microsoft.Net.Http" version="2.2.29" targetFramework="net452" />
   <package id="Newtonsoft.Json" version="7.0.1" targetFramework="net452" />
+  <package id="System.Interactive.Async" version="3.1.1" targetFramework="net452" />
   <package id="System.Management.Automation" version="6.1.7601.17515" targetFramework="net452" />
+  <package id="System.Net.Http" version="4.3.0" targetFramework="net452" />
   <package id="Zlib.Portable.Signed" version="1.11.0" targetFramework="net452" />
-</packages>
+</packages>
```
Comment (quoctruong) [inline Google.PowerShell/packages.config]: New line.

File: Google.PowerShell/BigQuery/BqTabledata.cs [hunk 8]
```diff
@@ -230,4 +231,206 @@ protected override void EndProcessing()
...
+            {
+                WriteError(new ErrorRecord(ex,
+                    $"Error while reading file {Filename}.",
+                    ErrorCategory.ReadError, Filename));
+                return;
+            }
+            var projectReference = new ProjectReference();
+            projectReference.ProjectId = InputObject.ProjectId;
+            var Client = new BigQueryClientImpl(projectReference, Service);
+            BigQueryJob bqj = null;
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: I prefer to leave this uninitialized. That way, I get a compile time check that bqj has been initialized in every case statement.

File: Google.PowerShell/BigQuery/BqTabledata.cs [hunk 9]
```diff
@@ -230,4 +231,206 @@ protected override void EndProcessing()
...
+            catch (Exception ex)
+            {
+                WriteError(new ErrorRecord(ex,
+                    $"Error while reading file {Filename}.",
+                    ErrorCategory.ReadError, Filename));
+                return;
+            }
+            var projectReference = new ProjectReference();
+            projectReference.ProjectId = InputObject.ProjectId;
+            var Client = new BigQueryClientImpl(projectReference, Service);
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Unless there is a specific reason you want to use the Service, the correct way to get a BigQueryClient is with `BigQueryClient.Create(string projectId)`.
If you do have to use the Service, there is a constructor `new BigQueryClientImpl(string projectId, Serivce)`. You do not need the ProjectReference object.

File: Google.PowerShell/BigQuery/BqTabledata.cs [hunk 10]
```diff
@@ -230,4 +231,206 @@ protected override void EndProcessing()
...
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        public int SkipLeadingRows { get; set; }
+
+        protected override void ProcessRecord()
+        {
+            Stream fileInput;
+            try
+            {
+                fileInput = new StreamReader(Filename).BaseStream;
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Streams should be controlled by a using statment. A stream should not be left unclosed.

```
using (Stream fileInput = File.OpenRead(Filename)) {
    ....
}
```

File: Google.PowerShell/BigQuery/BqTabledata.cs (hunk 8 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Hmm...  That's weird.  I had it like that earlier, but the compiler was still throwing a flag about it until I added the "=null".  It works now though, so something must have been misplaced.

File: Google.PowerShell/BigQuery/BqTabledata.cs (hunk 10 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Yes, of course.

File: Google.PowerShell/BigQuery/BqTabledata.cs (hunk 9 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Got it.  I was looking for a way to not need the ProjectReference.  I'll update my code in the add-tabledata cmdlet as well.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1 (hunk 1 above)
Comment (ahandley) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1]: WriteAppend and WriteTruncate don't throw errors naturally. Do we need tests using them on empty tables then or is it fine with just testing WriteIfEmpty?

File: Google.PowerShell/BigQuery/BqTabledata.cs (hunk 4 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/BqTabledata.cs]: Got them both now!

File: Google.PowerShell.IntegrationTests/BigQuery/classics_broken.csv (hunk 2 above)
Comment (ahandley) [inline Google.PowerShell.IntegrationTests/BigQuery/classics_broken.csv]: Added all the newlines!

**Comment by ahandley** [comment]
Fixed, PTAL.

**Comment by ahandley** [comment]
Thanks!
